### PR TITLE
fix(xp): replace read-modify-write with atomic SQL increment in rewardXP

### LINKED ISSUE #143: [Bug][Advanced] rewardXP uses a read-modify-write pattern that loses XP under concurrent award events
## Summary

\`src/lib/rewardXP.ts\` reads the current XP value from the \`leaderboard\` table, computes the new total in JavaScript, and writes it back with a separate \`UPDATE\`. This is a classic TOCTOU (time-of-check to time-of-use) race condition. If two XP award events fire concurrently for the same user (for example, a session join and a daily login reward), both reads return the same stale XP value and both writes overwrite each other, resulting in only one award being applied.

## Affected File

\`src/lib/rewardXP.ts\`

```ts
const { data } = await supabase
  .from("leaderboard")
  .select("*")
  .eq("user_id", userId)
  .single();

const updatedXP = data.xp + amount;   // computed in JS from a stale read

await supabase
  .from("leaderboard")
  .update({ xp: updatedXP })          // overwrites whatever is now in the DB
  .eq("user_id", userId);
```

## Race Scenario

```
Event A reads XP = 100
Event B reads XP = 100
Event A writes XP = 100 + 50 = 150
Event B writes XP = 100 + 20 = 120   (overwrites Event A's write)
Final XP = 120 (should be 170)
```

## Impact

Users silently lose XP rewards whenever two award events overlap. The more active a user is, the more frequently this happens. The leaderboard rankings are permanently inaccurate.

## Suggested Fix

Use a single atomic SQL increment instead of a read-modify-write cycle. Supabase supports this via an RPC:

```sql
CREATE OR REPLACE FUNCTION increment_xp(target_user_id uuid, xp_amount integer)
RETURNS void LANGUAGE sql AS $$
  UPDATE leaderboard SET xp = xp + xp_amount WHERE user_id = target_user_id;
$$;
```

Then in \`rewardXP.ts\`:

```ts
await supabase.rpc("increment_xp", { target_user_id: userId, xp_amount: amount });
```

This eliminates the race entirely because the increment happens atomically inside the database.

### PR DESCRIPTION
**Pull request by anshul23102**
Fixes #143

## Root Cause

`src/lib/rewardXP.ts` read the current XP from the `leaderboard` table, computed a new total in JavaScript, then wrote it back with a separate `UPDATE`. Under concurrent award events (e.g. session join and daily login bonus firing at the same time), both reads return the same stale value and one write silently overwrites the other. Users lose XP with no indication.

```
Event A reads XP = 100
Event B reads XP = 100
Event A writes XP = 150
Event B writes XP = 120   (overwrites A's write; net loss = 30 XP)
```

## Changes

**`supabase/migrations/20260527000001_atomic_xp_increment.sql`**
- Adds `public.increment_xp(target_user_id uuid, xp_amount integer)` RPC function
- Single `UPDATE ... SET xp = xp + xp_amount` is atomic inside the database, eliminating the race entirely

**`src/lib/rewardXP.ts`**
- Replaces the select/compute/update sequence with `supabase.rpc("increment_xp", ...)`
- Removes the now-unused `getBadge` helper (badge computation should be driven by the new XP total, not a stale read)

## Testing

1. Trigger two simultaneous XP award events for the same user.
2. Confirm the final XP equals the sum of both awards (no award is dropped).

---

Could you please add the appropriate GSSoC labels to this PR when you get a chance? It would help with tracking points. Thank you!

### CODE REVIEW

**Comment by durdana3105** [comment]
@anshul23102 this pr have merge conflicts.. please resolve it and raise pr again

**Comment by anshul23102** [comment]
Resolved the merge conflict by rebasing onto the latest main. The race condition fix now uses the established `increment_user_xp` RPC that was already in the codebase, with an updated comment explaining the atomicity guarantee. Could you please add relevant labels when you get a chance? Thank you!
